File: Solution TreeSet.cpp

#include <set>
#include <unordered_map>
using namespace std;

class NumberContainers {
    
    unordered_map<int, int> indexToValue;
    unordered_map<int, set<int>> valueToIndex;
    inline static const int NOT_FOUND = -1;

public:
    NumberContainers() = default;

    void change(int index, int number) {

        //C++20: indexToValue.contains(index). Done the old-fashioned way for compatibility.
        if (indexToValue.find(index) != indexToValue.end()) {

            int valueToBeReplaced = indexToValue[index];
            valueToIndex[valueToBeReplaced].erase(index);

            if (valueToIndex[valueToBeReplaced].empty()) {
                valueToIndex.erase(valueToBeReplaced);
            }
        }
        indexToValue[index] = number;
        valueToIndex[number].insert(index);
    }

    int find(int number) {
        return valueToIndex.find(number) != valueToIndex.end() ? *valueToIndex[number].begin() : NOT_FOUND;
    }
};

File: Solution TreeSet.java

import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;

public class NumberContainers {

    private static Map<Integer, Integer> indexToValue;
    private static Map<Integer, TreeSet<Integer>> valueToIndex;
    private static final int NOT_FOUND = -1;

    public NumberContainers() {
        indexToValue = new HashMap<>();
        valueToIndex = new HashMap<>();
    }

    public void change(int index, int number) {
        if (indexToValue.containsKey(index)) {

            int valueToBeReplaced = indexToValue.get(index);
            valueToIndex.get(valueToBeReplaced).remove(index);

            if (valueToIndex.get(valueToBeReplaced).isEmpty()) {
                valueToIndex.remove(valueToBeReplaced);
            }
        }
        indexToValue.put(index, number);
        valueToIndex.putIfAbsent(number, new TreeSet<>());
        valueToIndex.get(number).add(index);
    }

    public int find(int number) {
        return valueToIndex.containsKey(number) ? valueToIndex.get(number).first() : NOT_FOUND;
    }
}
